### PR TITLE
Fix Issue 19018 - lexer do not allow invalid integer literal

### PR DESCRIPTION
**Pull request by kubo39**
This is blocked by https://github.com/dlang/druntime/pull/2232

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
* Since druntime required changes this should be a deprecation
* Tests need to be added

File: test/compilable/fix19018.d [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/* REQUIRED_ARGS: -dw
```
Comment (wilzbach) [inline test/compilable/fix19018.d]: FWIW we typically use `-de` here and put the test in fail_compilation.
The advantage is that once the deprecation is over, the test doesn't need to be moved, but there's no strict rule on this.

File: src/dmd/lexer.d [hunk 2]
```diff
@@ -1955,6 +1969,11 @@ class Lexer : ErrorHandler
             error("integer overflow");
             err = true;
         }
+        if ((base == 2 && !anyBinaryDigitsUS) ||
+            (base == 16 && !anyHexDigitsNoSingleUS))
+        {
+            deprecation("Invalid integer literal is used, use valid one");
```
Comment (wilzbach) [inline src/dmd/lexer.d]: Maybe mention the name of the literal here?
E.g.

> `0x` isn't a valid integer literal

Comment (jacob-carlborg) [reply src/dmd/lexer.d]: Please include what literal to use to get the current behavior.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]
Please squash commits when done.

File: src/dmd/lexer.d [hunk 3]
```diff
@@ -1955,6 +1969,14 @@ class Lexer : ErrorHandler
             error("integer overflow");
             err = true;
         }
+        if ((base == 2 && !anyBinaryDigitsUS) ||
+            (base == 16 && !anyHexDigitsNoSingleUS))
+        {
+            auto tmp = new char[p-start+1];
+            tmp[0 .. p-start] = start[0 .. p-start];
+            tmp[$-1] = cast(char) null;
+            deprecation("%s isn't a valid integer literal", tmp.ptr);
```
Comment (JinShil) [inline src/dmd/lexer.d]: Code that is quoted in messages should be surrounded by tick marks.  In other words...

```
`%s` isn't a valid integer literal
```

**Review by JinShil — CHANGES_REQUESTED** [review]

**Comment by kubo39** [comment]
I changed commit message, should I change pull request title too?

**Comment by jacob-carlborg** [comment]
That's not necessary.

**Review by JinShil — APPROVED** [review]

**Review by jacob-carlborg — APPROVED** [review]

**Comment by kubo39** [comment]
Thanks for all reviews!

**Comment by schveiguy** [comment]
Am I missing something or why is `0b_` allowed?

**Pull request merged**